src/version.rs

```diff
@@ -178,6 +178,36 @@ pub fn upgrade_requirement(req: &str, version: &semver::Version) -> Result<Optio
     }
 }
 
+/// Upgrade an existing requirement to a new version if `preserve_precision` is set; otherwise,
+/// return the new version
+pub(crate) fn upgrade_requirement_conditionally(
+    req: &str,
+    version: &str,
+    preserve_precision: bool,
+) -> Result<Option<String>> {
+    if preserve_precision {
+        let version = semver::Version::parse(version)?;
+        upgrade_requirement(req, &version)
+    } else {
+        Ok(Some(version.to_owned()))
+    }
+}
+
+/// Upgrade an existing requirement to a new version if `preserve_precision` is set; otherwise,
+/// return the new version. If no upgrade is necessary, return the old requirement.
+pub(crate) fn upgrade_requirement_with_default(
+    req: Option<&str>,
+    version: &str,
+    preserve_precision: bool,
+) -> Result<String> {
+    if let Some(req) = req {
+        upgrade_requirement_conditionally(req, version, preserve_precision)
+            .map(|new_req| new_req.unwrap_or_else(|| req.to_owned()))
+    } else {
+        Ok(version.to_owned())
+    }
+}
```

Comment:
These are fairly specialized functions, let's keep them as close to their use as possible and with the most limited reasonable visibility.

Reply:
In my latest revision, these functions are gone.